Update Mailbox.php - correct mb_convert_encoding() error

trianity opened the pull request:
The handle unsupported encoding we need add one more check with mb_list_encodings(), that will return as an array the supported encodings in the PHP instance. And if the $fromEncoding value is not on the list, the encoding will be untouched.

See: https://github.com/mautic/mautic/issues/14468

<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 5.x)
* a.b for any bug fixes (e.g. 4.4, 5.1)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the 5.2 branch)          | 🟢 <!-- Use emojis to indicate positive (green) or negative (red) for each item in the table. -->
| New feature/enhancement? (use the a.x branch)      | 🔴
| Deprecations?                          | 🔴
| BC breaks? (use the c.x branch)        | 🔴
| Automated tests included?              | 🟢 <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/user-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation-new#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

## Description

The error occurred while running console command `mautic:email:fetch` when the monitored email box contained unsupported text encoding in the email text.

**The issue prevents the console command `mautic:email:fetch` from running, so it is considered a critical error.**

According the manual page of mb_convert_encoding() this is an old issue with the function, and the @ operator is not enough the handle.

The original function in MailBox.php about line 1051, that fails:

```
protected function convertStringEncoding($string, $fromEncoding, $toEncoding)
    {
        $convertedString = null;
        if ($string && $fromEncoding != $toEncoding) {
            $convertedString = @iconv($fromEncoding, $toEncoding.'//IGNORE', $string);
            if (!$convertedString && extension_loaded('mbstring')) {
                $convertedString = @mb_convert_encoding($string, $toEncoding, $fromEncoding);
            }
        }

        return $convertedString ?: $string;
    }
```

The handle unsupported encoding we need add one more check with mb_list_encodings(), that will return as an array the supported encodings in the PHP instance. And the $fromEncoding is not on the list, it will be untouched.

```
protected function convertStringEncoding($string, $fromEncoding, $toEncoding)
    {
        $convertedString = null;
        if ($string && $fromEncoding != $toEncoding) {
            $convertedString = @iconv($fromEncoding, $toEncoding.'//IGNORE', $string);
            if (!$convertedString && extension_loaded('mbstring')) {
                $listOfEncodings = mb_list_encodings();
                if (in_array($fromEncoding, $listOfEncodings)) {
                    $convertedString = @mb_convert_encoding($string, $toEncoding, $fromEncoding);
                }
            }
        }

        return $convertedString ?: $string;
    }
```

The suggested solution will handle the bug fully.


---
### 📋 Steps to test this PR:

<!--
This part is crucial. Take the time to write very clear, annotated and step by step test instructions, because testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. 

<!--
If you have any deprecations and backwards compatibility breaks, list them here along with the new alternative.
-->

trianity committed:
Update Mailbox.php - correct mb_convert_encoding() error

The handle unsupported encoding we need add one more check with mb_list_encodings(), that will return as an array the supported encodings in the PHP instance. And if the $fromEncoding value is not on the list, the encoding will be untouched.

trianity commented:
@matbcvo Hello, Here is the new request.

matbcvo commented:
@trianity Could you cover this code change with a PHPUnit test?

File: `app/bundles/EmailBundle/Tests/MonitoredEmail/MailboxTest.php`

https://github.com/mautic/mautic/blob/5.2/app/bundles/EmailBundle/Tests/MonitoredEmail/MailboxTest.php

You can use the new methods provided below and make adjustments as needed.

```php
use Mautic\EmailBundle\MonitoredEmail\Mailbox;
use PHPUnit\Framework\TestCase;


class MailboxTest extends TestCase
{
    // ... other methods

    public function testEncodingConversionWithValidEncoding(): void
    {
        $string = "Hello, world!";
        $fromEncoding = "UTF-8";
        $toEncoding = "ISO-8859-1";

        // Mock dependencies
        $parametersHelper = $this->getMockBuilder(CoreParametersHelper::class)
            ->disableOriginalConstructor()
            ->getMock();
        $pathsHelper = $this->getMockBuilder(PathsHelper::class)
            ->disableOriginalConstructor()
            ->getMock();

        $mailbox = new Mailbox($parametersHelper, $pathsHelper);

        $result = $this->invokeConvertStringEncoding($mailbox, $string, $fromEncoding, $toEncoding);

        $this->assertNotSame($string, $result, "Converted string should differ from the original string.");
    }

    public function testEncodingConversionWithInvalidEncoding(): void
    {
        $string = "Hello, world!";
        $fromEncoding = "INVALID-ENC";
        $toEncoding = "ISO-8859-1";

        // Mock dependencies
        $parametersHelper = $this->getMockBuilder(CoreParametersHelper::class)
            ->disableOriginalConstructor()
            ->getMock();
        $pathsHelper = $this->getMockBuilder(PathsHelper::class)
            ->disableOriginalConstructor()
            ->getMock();

        $mailbox = new Mailbox($parametersHelper, $pathsHelper);

        $result = $this->invokeConvertStringEncoding($mailbox, $string, $fromEncoding, $toEncoding);

        $this->assertSame($string, $result, "Conversion with invalid encodings should return the original string.");
    }

    public function testEncodingConversionFallback(): void
    {
        if (!extension_loaded('mbstring')) {
            $this->markTestSkipped("Test skipped as the mbstring extension is not installed.");
        }

        $string = "Hello, world!";
        $fromEncoding = "UTF-8";
        $toEncoding = "ISO-8859-1";

        // Mock dependencies
        $parametersHelper = $this->getMockBuilder(CoreParametersHelper::class)
            ->disableOriginalConstructor()
            ->getMock();
        $pathsHelper = $this->getMockBuilder(PathsHelper::class)
            ->disableOriginalConstructor()
            ->getMock();

        $mailbox = new Mailbox($parametersHelper, $pathsHelper);

        $listEncodings = mb_list_encodings();
        if (in_array($fromEncoding, $listEncodings)) {
            $result = $this->invokeConvertStringEncoding($mailbox, $string, $fromEncoding, $toEncoding);
            $this->assertNotSame($string, $result, "Converted string should differ from the original string if conversion succeeds.");
        } else {
            $this->markTestSkipped("Test skipped as the encoding is not supported by the current mbstring configuration.");
        }
    }


    private function invokeConvertStringEncoding(Mailbox $mailbox, string $string, string $fromEncoding, string $toEncoding): string
    {
        $reflection = new \ReflectionClass($mailbox);
        $method = $reflection->getMethod('convertStringEncoding');
        $method->setAccessible(true);

        return $method->invokeArgs($mailbox, [$string, $fromEncoding, $toEncoding]);
    }
}
```

Don't forget to run PHP CS fixer on it.
```
composer fixcs -- app/bundles/EmailBundle/Tests/MonitoredEmail/MailboxTest.php
```

trianity commented:
> @trianity Could you cover this code change with a PHPUnit test?
> 
> File: `app/bundles/EmailBundle/Tests/MonitoredEmail/MailboxTest.php`
> 
> https://github.com/mautic/mautic/blob/5.2/app/bundles/EmailBundle/Tests/MonitoredEmail/MailboxTest.php
> 

@matbcvo Of course, but I am sorry, I will do it a little bit later ...

escopecz reviewed (commented):
Do you think you could write a test that will show the issue and pass with your changes?

trianity commented:
> Do you think you could write a test that will show the issue and pass with your changes?

Yes, but have some time issues :-)

trianity committed:
Merge branch 'mautic:5.2' into 5.2

trianity committed:
Update MailboxTest.php

Cover the code change in app/bundles/EmailBundle/MonitoredEmail/Mailbox.php with a PHPUnit test for convertStringEncoding()

trianity commented:
@escopecz  @matbcvo The PHPUnit test is added to the function change with 3 data set in one test.

escopecz commented on app/bundles/EmailBundle/Tests/MonitoredEmail/MailboxTest.php:
```diff
@@ -184,4 +184,41 @@ public function testUseAttachments(): void
         new \Mautic\EmailBundle\MonitoredEmail\Mailbox($parametersHelper, $pathsHelper);
     }
+
+    public function testConvertStringEncoding(): void
+    {
+        $mailbox = $this->getMockBuilder(\Mautic\EmailBundle\MonitoredEmail\Mailbox::class)
+                        ->disableOriginalConstructor()
+                        ->getMock();
+
+        $reflection = new \ReflectionClass(\Mautic\EmailBundle\MonitoredEmail\Mailbox::class);
+        $method     = $reflection->getMethod('convertStringEncoding');
+        $method->setAccessible(true);
+
+        $this->assertTrue(extension_loaded('mbstring'), 'mbstring extension is loaded');
+
+        // Prepare 1. test data: valid fromEncoding
+        $string       = 'some UTF8 text with öüóőúéáűíä and ÖÜÓŐÚÉÁŰÍßÄŁ';
+        $fromEncoding = 'UTF-8';
+        $toEncoding   = 'ISO-8859-1';
+        $result       = $method->invoke($mailbox, $string, $fromEncoding, $toEncoding);
+
+        // Assert the expected outcome
+        $this->assertNotNull($result);
+        $this->assertNotSame($string, $result, 'Converted string should differ from the original string.');
+
+        // Prepare 2. test data: fromEncoding not differ from toEncoding
+        $fromEncoding2 = $toEncoding2 = 'UTF-8';
+        $result2       = $method->invoke($mailbox, $string, $fromEncoding2, $toEncoding2);
+        $this->assertNotNull($result);
+        $this->assertSame($string, $result2, 'The string should not differ from the original string.');
+
+        // Prepare 3. test data: with invalid fromEncoding
+        $string3       = 'some text with special chars: öüóőúéáűíä and ÖÜÓŐÚÉÁŰÍßÄŁ';
+        $fromEncoding3 = 'unicode-1-1-utf'; // invalid fromEncoding
+        $toEncoding3   = 'UTF-8';
+        $result3       = $method->invoke($mailbox, $string3, $fromEncoding3, $toEncoding3);
+        $this->assertNotNull($result);
+        $this->assertSame($string3, $result3, 'The string should not differ from the original string.');
+    }
```
This may be easier to read:
```
/**
     * @dataProvider testConvertStringEncodingDataProvider
     */
    public function testConvertStringEncoding(string $providedText, string $fromEncoding, string $toEncoding, string $expectedText): void
    {
        $mailbox    = $this->createMock(\Mautic\EmailBundle\MonitoredEmail\Mailbox::class);
        $reflection = new \ReflectionClass(\Mautic\EmailBundle\MonitoredEmail\Mailbox::class);
        $method     = $reflection->getMethod('convertStringEncoding');
        $method->setAccessible(true);

        $result = $method->invoke($mailbox, $providedText, $fromEncoding, $toEncoding);

        $this->assertNotNull($result);
        $this->assertSame($expectedText, $result);
    }

    public function testConvertStringEncodingDataProvider(): Iterator
    {
        yield 'valid fromEncoding' => [
            'providedText' => 'some UTF8 text with öüóőúéáűíä and ÖÜÓŐÚÉÁŰÍßÄŁ',
            'fromEncoding' => 'UTF-8',
            'toEncoding'   => 'ISO-8859-1',
            'expectedText' => 'some UTF8 text with öüóőúéáűíä and ÖÜÓŐÚÉÁŰÍßÄŁ',
        ];

        yield 'fromEncoding not differ from toEncoding' => [
            'providedText' => 'some UTF8 text with öüóőúéáűíä and ÖÜÓŐÚÉÁŰÍßÄŁ',
            'fromEncoding' => 'UTF-8',
            'toEncoding'   => 'UTF-8',
            'expectedText' => 'some UTF8 text with öüóőúéáűíä and ÖÜÓŐÚÉÁŰÍßÄŁ',
        ];

        yield 'with invalid fromEncoding' => [
            'providedText' => 'some text with special chars: öüóőúéáűíä and ÖÜÓŐÚÉÁŰÍßÄŁ',
            'fromEncoding' => 'unicode-1-1-utf', // invalid fromEncoding
            'toEncoding'   => 'UTF-8',
            'expectedText' => 'some UTF8 text with öüóőúéáűíä and ÖÜÓŐÚÉÁŰÍßÄŁ',
        ];
    }
```
(untested, just pseudocode showing the idea)

trianity replied:
Completed the test with independent functions, but using dataset also a good idea.

escopecz replied:
Any idea why does the CI marked the newly added condition as not being executed by the tests? Can you step over it with xdebug to confirm?

matbcvo replied:
I did some debugging and noticed that `mb_convert_encoding` is not executed on the given string `some text with non-ascii chars öüóőúéáűíä and ÖÜÓŐÚÉÁŰÍßÄŁ`.

To ensure `mb_convert_encoding` is executed, I suggest using the following string:
```
$string = "こんにちは"; // "Hello" in Japanese
$fromEncoding = "UTF-8";
$toEncoding = "ISO-8859-1";
```
Or add some Japanese characters to the original string to trigger it.

matbcvo replied:
Hello @trianity, could you take a look at the PHPUnit test issue described above?

trianity committed:
Update MailboxTest.php with convertStringEncoding() tests

Improved tests for all cases: convertStringEncoding() from app/bundles/EmailBundle/MonitoredEmail/Mailbox.php

trianity committed:
Update MailboxTest.php

Adding missing use statement of: Mautic\EmailBundle\MonitoredEmail\Mailbox;

trianity committed:
Update MailboxTest.php style fix

Correct \Mautic\EmailBundle\MonitoredEmail\Mailbox:class to Mailbox:class as the use statement is in the class.

RCheesley commented:
Removing from the milestone due to unresolved feedback.

RCheesley committed:
Merge branch '5.2' into 5.2

escopecz reviewed (approved):
It looks like this is good to go 👍

escopecz commented:
Oh, right. This must be rebased on top of the 6.0 branch as Mautic 5 is in the "security fixes only" mode.

https://mautic.org/releases/

RCheesley commented:
@all-contributors please add @stone-creator for usertesting
